feat(reconciliation): Do hybrid reconciliation

fredrb opened the pull request:
## Description

This is a PR attempting to fix the reconciliation by still keeping deletes in central, but sending a set of `unchangedIDs` to central at the `SYNC` message, as a batch `SYNC` for any resource that hasn't changed. 

A more detailed description of the problem and solutions can be found [here](https://docs.google.com/document/d/1m9xgPAg9-Q_ukcAKxbHH11_oKl-__qQCb2CJ_NvMGKc/edit?usp=sharing).

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~~Evaluated and added CHANGELOG entry if required~~
- [ ] ~~Determined and documented upgrade steps~~
- [ ] ~~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

### Here I tell how I validated my change

Turned the feature flag on, and let CI run with new reconciliation. Given that this PR makes use of Central reconciliation, all reconciliation related tests shall serve as a regression test.

- [x] CI 
    - [x] ReconciliationTest
    - [x] UpgradeTest (backup recover)

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

vikin91 commented:
@fredrb is this ready for a review? If yes, then please fill the `Here I tell how I validated my change`, as I would like to follow the idea for solution while looking at the code.

vikin91 commented on sensor/common/sensor/central_sender_impl.go:
```diff
@@ -112,10 +114,10 @@ func (s *centralSenderImpl) send(stream central.SensorService_CommunicateClient,
 
 			if msg.GetEvent().GetSynced() != nil {
-				log.Info("Sending synced signal to Central")
-				if s.onSync != nil {
-					go s.onSync()
```
By looking at this diff, the `onSync` function seems not to be used anymore. Could we remove it maybe?

vikin91 reviewed (commented):
The PR looks really well! Thanks for explaining me the details of this idea on Slack and in the doc. 

I am leaving few comments to the code but none of them are blockers.
What bothers me a bit are the :x: from the CI. I would like to wait with reviewing ✅ until they are resolved.

fredrb commented:
> What bothers me a bit are the ❌ from the CI

@vikin91 for some reason `s390x` build failed on my PR, but I couldn't figure out why. This caused images to not be pushed and thus prow being entirely red. 

I have rebased with master and fixed your comments. If this build fails again I'll investigate further.

vikin91 commented on sensor/common/deduper/unchanged.go:
```diff
@@ -7,26 +7,26 @@ import (
 
-// ObservationSet holds IDs that were seen during the initial sensor sync.
-type ObservationSet struct {
+// ClosableSet holds IDs that were seen during the initial sensor sync.
```
```suggestion
// ClosableSet is a set of strings. Items can be added to the set as long as the set is open. Closing the set ignores all future `Add` operations but allows to fetch already stored items. 
```

fredrb replied:
Done. Just changed to also add info about it being concurrency-safe.

vikin91 commented on sensor/common/deduper/unchanged.go:
```diff
@@ -7,26 +7,26 @@ import (
 
-// NewObservationSet creates an empty sync observation set.
-func NewObservationSet() *ObservationSet {
+// NewClosableSet creates an empty sync observation set.
```
Update comment pls.

vikin91 commented on sensor/common/deduper/unchanged.go:
```diff
@@ -7,26 +7,26 @@ import (
 
-// LogObserved parses a key `k` and adds to the observation set.
-func (s *ObservationSet) LogObserved(stringKey string) {
+// AddIfOpen parses a key `k` and adds to the observation set.
```
```suggestion
// AddIfOpen adds item to an unclosed set. If set is closed, this operation does nothing
```

vikin91 commented on sensor/common/deduper/unchanged.go:
```diff
@@ -35,7 +35,7 @@ func (s *ObservationSet) LogObserved(stringKey string) {
 }
 
 // Close will stop processing of any further keys and return a list of parsed keys.
```
```suggestion
// Close marks set as closed preventing any new items to be added. It returns the currently stored items.
```

vikin91 reviewed (commented):
I think, the doc-comments to the ClosableSet could be refreshed a bit.

vikin91 reviewed (approved):
Thanks, it looks good! The IBM build is green now, so let's wait for the CI.
You can also polish the comments for the `ClosableSet` if there is an opportunity for pushing another commit.

lvalerom commented on sensor/common/deduper/deduper.go:
```diff
@@ -58,24 +61,35 @@ var (
 	}
 }
 
+func buildKey(typ, id string) string {
+	return fmt.Sprintf("%s:%s", typ, id)
+}
+
+func getKey(msg *central.MsgFromSensor) string {
+	event := msg.GetEvent()
+	return buildKey(eventPkg.GetEventTypeWithoutPrefix(event.GetResource()), event.GetId())
+}
```
[nit] This functions are also defined [here](https://github.com/stackrox/stackrox/blob/4ba873b685b734595366b78880c6c2af8ada9e9d/central/hash/manager/deduper.go#L124-L131). Would it make sense to maybe extract them in a common package under `pkg`?

fredrb replied:
Created a follow-up for this: [ROX-20693](https://issues.redhat.com/browse/ROX-20693)
I think there might be other things we can pull up to a shared package.

lvalerom reviewed (approved):
Looks good! Sorry for the late review 😅

fredrb commented:
/test ?

fredrb commented:
- `gke-ui-e2e-tests` seem to be failing for every PR.
- `gke-version-compatibility-tests` failed on the post test but all tests seem to have passed. I will re-run them just in case.

fredrb commented:
/test gke-version-compatibility-tests

fredrb commented:
/test gke-version-compatibility-tests

fredrb commented:
/test gke-version-compatibility-tests
/test ocp-4-10-qa-e2e-tests

connorgorman commented on central/sensor/service/connection/sensorevents.go:
```diff
@@ -87,10 +96,19 @@ func (s *sensorEventHandler) addMultiplexed(ctx context.Context, msg *central.Ms
-			log.Infof("Ignoring SYNC from cluster %s (ClusterID:%s) since reconciliationMap is already closed", s.cluster.GetName(), s.cluster.GetId())
-			return
+
+		unchangedIDs := event.GetSynced().GetUnchangedIds()
```
How much can this message be realistically?

fredrb replied:
note: fallback to classic reconciliation if max message receive is received.

fredrb replied:
Created a task to address the fallback to classic reconciliation [ROX-20878](https://issues.redhat.com/browse/ROX-20878).

connorgorman commented on pkg/deduperkey/key.go:
```diff
@@ -41,6 +44,25 @@ type Key struct {
+// ParseKeySlice returns a list of Key objects from a list a string formatted keys. An error returned means that some
+// of the keys might have failed when being parsed.
+func ParseKeySlice(keys []string) ([]Key, error) {
+	if keys == nil {
```
This check isn't necessary so I think it's better to omit it

fredrb replied:
Done

connorgorman commented on proto/internalapi/central/sensor_events.proto:
```diff
@@ -63,7 +63,9 @@ message SensorEvent {
 
-    message ResourcesSynced {}
+    message ResourcesSynced {
+        repeated string unchanged_ids = 1;
```
My only concern is about the size of this message

fredrb replied:
See comment above about problem doc and the new task to implement classic reconciliation fallback.

connorgorman commented on sensor/common/deduper/unchanged.go:
```diff
@@ -0,0 +1,45 @@
+
+// ClosableSet is a concurrency-safe set of strings. Items can be added to the set as long as the set is open.
+// Closing the set ignores all future `Add` operations but allows to fetch already stored items.
+type ClosableSet struct {
```
I'm confused by this a bit. Why would you ever add anything to a closed set? Doesn't that imply a timing bug where a synced resource is trying to be added after close?

fredrb replied:
>  Why would you ever add anything to a closed set?

We shouldn't ever. That's why the set is being closed after the reconciliation finishes.

> Doesn't that imply a timing bug where a synced resource is trying to be added after close?

It does, because the set should be closed as soon as sensor observes a `ResourcesSynced` event. If any sensor event gets triggered with `ResourceAction_SYNC_RESOURCE` it's a timing issue.

But my decision was to ignore that event rather than log an error. I will change this a bit to show an error if this case happens.

fredrb replied:
I've changed the code and ended up removing this struct all together. I think it's simpler now.

connorgorman commented on sensor/common/deduper/unchanged.go:
```diff
@@ -0,0 +1,45 @@
+
+// AddIfOpen adds item to an unclosed set. If set is closed, this operation does nothing.
+func (s *ClosableSet) AddIfOpen(stringKey string) {
+	if s.open.Load() {
```
nit: I find it cleaner to check the opposite cases and then return which avoids extra indentation

fredrb replied:
Done

connorgorman commented on sensor/debugger/central/communicate_server.go:
```diff
@@ -73,6 +78,8 @@ func MakeFakeCentralWithInitialMessages(initialMessages ...*central.MsgToSensor)
 		messageCallbackLock:  sync.RWMutex{},
 		centralStubMessagesC: make(chan *central.MsgToSensor, 1),
+		deduperState:         &central.DeduperState{ResourceHashes: make(map[string]uint64)},
+		deduperStateLock:     sync.RWMutex{},
```
No need to initialize a mutex

fredrb replied:
Done

connorgorman commented on sensor/common/sensor/central_sender_impl.go:
```diff
@@ -112,9 +110,20 @@ func (s *centralSenderImpl) send(stream central.SensorService_CommunicateClient,
-					go s.onSync()
+				unchangedIds := s.observationSet.Close()
+				// Enhance sync with all the observed IDs
+				msg.GetEvent().GetSynced().UnchangedIds = unchangedIds
```
This is a strange usage of the getters. It's cleaner if you modify this in the if above as in

```
if syncEvent := msg.GetEvent().GetSynced(); syncEvent != nil 
```

then you operate on syncEvent

fredrb replied:
Done

connorgorman commented on sensor/common/sensor/central_sender_impl.go:
```diff
@@ -112,9 +110,20 @@ func (s *centralSenderImpl) send(stream central.SensorService_CommunicateClient,
+
+				// Set metrics count for unchanged IDs
+				for i, key := range unchangedIds {
+					parts := strings.SplitN(key, ":", 2)
```
This section seems unnecessary? just inc by the len(of unchanged ids) and print out the ideas using %+v?

fredrb replied:
Removed

connorgorman commented on sensor/common/deduper/deduper.go:
```diff
@@ -58,24 +61,35 @@ var (
-	lastSent map[Key]uint64
+	stream         messagestream.SensorMessageStream
+	lastSent       map[deduperkey.Key]uint64
+	observationSet *ClosableSet
```
This is a mega nit so feel free to ignore, but I think it should be Closeable

fredrb replied:
Done

connorgorman reviewed (commented):
I added some comments, most of which will not impact the code though please respond to some of the conversations before merging

fredrb commented on pkg/features/list.go:
```diff
@@ -65,7 +65,7 @@ var (
 
 	// SensorReconciliationOnReconnect enables sensors to support reconciliation when reconnecting
-	SensorReconciliationOnReconnect = registerFeature("Enable Sensors to support reconciliation on reconnect", "ROX_SENSOR_RECONCILIATION", false)
+	SensorReconciliationOnReconnect = registerFeature("Enable Sensors to support reconciliation on reconnect", "ROX_SENSOR_RECONCILIATION", true)
```
Switch to false
